noir_stdlib/src/collections/bounded_vec.nr

```diff
@@ -93,3 +95,64 @@ impl<T, MaxLen> BoundedVec<T, MaxLen> {
         ret
     }
 }
+
+impl<T, MaxLen> Eq for BoundedVec<T, MaxLen> where T: Eq {
+    fn eq(self, other: BoundedVec<T, MaxLen>) -> bool {
+        let equal_lens = self.len == other.len;
+
+        let mut ret = true;
+        let mut exceeded_len = false;
+        for i in 0..MaxLen {
+            exceeded_len |= i == self.len;
+            if !exceeded_len {
+                ret &= (self.storage[i] == other.storage[i]);
+            }
+        }
+        equal_lens & ret
```

Comment:
```suggestion
        let mut ret = self.len == other.len;
        let mut exceeded_len = false;
        for i in 0..MaxLen {
            exceeded_len |= i == self.len;
            if !exceeded_len {
                ret &= self.storage[i] == other.storage[i];
            }
        }
        ret
```
I'm assuming equalities are expected to be cheaper than comparisons? I wonder what the constraint difference would be if we did:
```rs
        let mut ret = self.len == other.len;
        for i in 0..MaxLen {
            if i >= self.len {
                ret &= self.storage[i] == other.storage[i];
            }
        }
        ret
```

Reply:
Nice to know to lean towards equality in the future even if the difference isn't too great

Reply:
We can actually improve this with

```rust
let mut ret = vec1.len == vec2.len;
let mut not_exceeded_len = true;
for i in 0..5 {
    if not_exceeded_len {
        ret &= vec1.storage[i] == vec2.storage[i];
    }
    not_exceeded_len &= i != vec1.len;
}
ret
```

which takes us down to 

```
+---------+----------+----------------------+--------------+----------------------+
| Package | Function | Expression Width     | ACIR Opcodes | Backend Circuit Size |
+---------+----------+----------------------+--------------+----------------------+
| temp    | main     | Bounded { width: 3 } | 68           | 2847                 |
+---------+----------+----------------------+--------------+----------------------+
```

Reply:
The difference there being using `&` instead of `|` for `not_exceeded_len`?
The `!` operation was also moved from a raw `!var` to a combined `!=` so that could be it as well.

Reply:
We're also not handling empty `BoundedVec`s properly, if we fix that then we lose the gains.

Reply:
Why aren't empty vecs being handled properly?

Reply:
Sorry, yeah "empty vec" was pretty vague.

Reply:
Once we can encapsulate the inner state of the `BoundedVec` so users can't fill the storage with garbage then we can make stronger assumptions about the end of the storage array.

Reply:
Should we consider that users can give an invalid state at all right now though? Since they can also do `vec.len = 1000;` on a vec with a max len of 10. I think we should just ignore such cases until we can prevent them with private fields.

Reply:
hmm, that's true. In that case we can just replace this implementation with a direct equality on the storage vector.